### ISSUE TITLE
define-syntax-class uses the wrong scopes ?

### ISSUE THREAD

**Issue by jsmaniac** [racket/racket#1483]
This simple program uses a macro `(my-define-syntax-class . opts)` which expands to `(define-syntax-class . opts)`. The defined name is not recognized as a syntax class, though.

```
syntax-parser: not defined as syntax class in: cls1
```

```
#lang racket
(require syntax/parse
         (for-syntax syntax/parse)
         (for-meta 2 racket/base syntax/parse))
(begin-for-syntax
  (define-syntax-rule (my-define-syntax-class . opts)
    (define-syntax-class . opts))

  (my-define-syntax-class cls1 (pattern _))
  (define-syntax-class cls2 (pattern _)))

(define-syntax foo
  (syntax-parser
    [:cls1 #'ok]
    [:cls2 #'ok]))
```

What solved the problem was to remove the use-site scope that macro-expansion otherwise automatically adds. The `remove-use-site-scope` function from my toolkit is based on `syntax-local-identifier-as-binding`, but also works for non-identifiers. The code below compiles and runs fine:

```
#lang racket
(require syntax/parse
         (for-syntax syntax/parse)
         (for-meta 2 racket/base syntax/parse phc-toolkit/untyped))
(begin-for-syntax
  (define-syntax (my-define-syntax-class stx)
    (syntax-case stx ()
      [(_ . opts)
       (remove-use-site-scope
        (datum->syntax #'name `(define-syntax-class . ,#'opts)))]))

  (my-define-syntax-class cls1 (pattern _))
  (define-syntax-class cls2 (pattern _)))

(define-syntax foo
  (syntax-parser
    [:cls1 #'ok]
    [:cls2 #'ok]))
```

I think `define-syntax-class` should use the scopes from the class name, and not the scope from the whole stx expression. A first guess would be that there is a `(datum->syntax stx …)` somewhere in the code for `define-syntax-class`, and it should be `(datum->syntax #'name …)` or something similar.

**Comment by rmculpepper** [comment]
It appears to be a general hygiene issue, not a syntax-parse issue, as demonstrated by the following program:

    #lang racket
    (require (for-syntax racket/match))
    (begin-for-syntax
      (define-syntax-rule (my-struct . opts)
        (struct . opts))
    
      (my-struct A (x y))
      (struct B (x y)))
    
    (define-syntax (foo stx)
      (match (A 1 2)
        [(A x y) #`(quote #,(list x y))]
        [(B x y) #`(quote #,(list x y))]))

which raises the error `match: syntax error in pattern in: (A x y)`. That is, the `syntax-local-value` call on `A` in the match pattern fails for some reason---presumably the same reason it fails when looking up the syntax class `cls1` in the original example.

**Comment by jsmaniac** [comment]
Hm. Just to be sure: Is it something I'm missing, or is it intended to be that way? I have the feeling that both examples should work without the `remove-use-site-scope` trick.

**Comment by rmculpepper** [comment]
Both programs work in Racket 6.1.

**Comment by jsmaniac** [comment]
[syntax-local-identifier-as-binding](http://docs.racket-lang.org/reference/stxtrans.html#%28def._%28%28quote._~23~25kernel%29._syntax-local-identifier-as-binding%29%29) was added in v6.3, which is when the new macro expander from @mflatt was used. Maybe that's just a limitation / new way of working of the new expander.

**Comment by sorawee** [comment]
This appears to be fixed in Racket 7.7 (the earliest Racket version I have). Closing, but feel free to reopen if any issue remains.
